extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableAdapter.java

```diff
@@ -228,231 +229,164 @@ public synchronized Optional<Schema> schema(final int schemaId) {
     }
 
     /**
-     * Return {@link TableDefinition table definition}.
-     *
-     * @return The table definition
-     */
-    public TableDefinition definition() {
-        // Load the table from the catalog.
-        return definition(null);
-    }
-
-    /**
-     * Return {@link TableDefinition table definition} with optional instructions for customizations while reading.
-     *
-     * @param instructions The instructions for customizations while reading (or null for default instructions)
-     * @return The table definition
+     * Retrieves the appropriate {@link Snapshot} based on the provided {@link IcebergReadInstructions}, or {@code null}
+     * if no snapshot is provided.
      */
-    public TableDefinition definition(@Nullable final IcebergInstructions instructions) {
-        // Load the table from the catalog.
-        return definition(null, instructions);
+    @Nullable
+    private Snapshot getSnapshot(final IcebergReadInstructions readInstructions) {
+        final Snapshot snapshot;
+        if (readInstructions.snapshot().isPresent()) {
+            snapshot = readInstructions.snapshot().get();
+        } else if (readInstructions.tableSnapshotId().isPresent()) {
+            snapshot = snapshot(readInstructions.tableSnapshotId().getAsLong())
+                    .orElseThrow(() -> new IllegalArgumentException(
+                            "Snapshot with id " + readInstructions.tableSnapshotId().getAsLong() + " not found for " +
+                                    "table " + tableIdentifier));
+        } else {
+            snapshot = null;
+        }
+        return snapshot;
     }
 
     /**
-     * Return {@link TableDefinition table definition} for the Iceberg table and snapshot id, with optional instructions
-     * for customizations while reading.
-     *
-     * @param snapshotId The identifier of the snapshot to load
-     * @param instructions The instructions for customizations while reading (or null for default instructions)
-     * @return The table definition
+     * Used to hold return value for {@link #getSpecAndSchema(IcebergReadInstructions)}.
      */
-    public TableDefinition definition(
-            final long snapshotId,
-            @Nullable final IcebergInstructions instructions) {
-        // Find the snapshot with the given snapshot id
-        final Snapshot tableSnapshot =
-                snapshot(snapshotId).orElseThrow(() -> new IllegalArgumentException(
-                        "Snapshot with id " + snapshotId + " not found for table " + tableIdentifier));
-
-        // Load the table from the catalog.
-        return definition(tableSnapshot, instructions);
+    private static final class SpecAndSchema {
```

Comment:
I might argue that `Snapshot` should be represented here rather than a re-written instruction object.

Reply:
Actually, maybe not. Rewriting might be exactly what we want.  @devinrsmith wants to include Schema instructions in these calls so maybe populating the instructions with snapshot now (and Schema later) is the right idea.